bluepill/tests/BPPackerTests.m

```diff
@@ -83,6 +83,51 @@ - (void)testPackingProvidesBalancedBundles {
     }
 }
 
+- (void)testSmartPackIfJsonFound {
+    self.config.testTimeEstimatesJsonFile = [BPTestHelper sampleTimesJsonPath];
+    self.config.testBundlePath = [BPTestHelper sampleAppBalancingTestsBundlePath];
+    self.config.testCasesToSkip = @[@"BPSampleAppTests/testCase000"];
+    self.config.numSims = @8;
+    BPApp *app = [BPApp appWithConfig:self.config withError:nil];
+    XCTAssert(app != nil);
+    NSArray<BPXCTestFile *> *bundles;
+    NSError *error;
+    bundles = [BPPacker packTests:app.testBundles configuration:self.config andError:&error];
+    XCTAssert(error ==  nil);
+    XCTAssert([app.testBundles count] == [bundles count]);
+}
+
+- (void)testSmartPackIfJsonMissing {
+    self.config.testTimeEstimatesJsonFile = @"invalid/times/file/path.json";
+    self.config.testBundlePath = [BPTestHelper sampleAppBalancingTestsBundlePath];
+    self.config.testCasesToSkip = @[@"BPSampleAppTests/testCase000"];
+    self.config.numSims = @8;
+    BPApp *app = [BPApp appWithConfig:self.config withError:nil];
+    XCTAssert(app != nil);
+    NSArray<BPXCTestFile *> *bundles;
+    NSError *error;
+    bundles = [BPPacker packTests:app.testBundles configuration:self.config andError:&error];
+    XCTAssert(error !=  nil);
+    XCTAssert([bundles count] == 0);
+}
+
+- (void)testSortByTimeEstimates {
+    self.config.testTimeEstimatesJsonFile = [BPTestHelper sampleTimesJsonPath];
+    self.config.testBundlePath = nil;
+    BPApp *app = [BPApp appWithConfig:self.config withError:nil];
+    XCTAssert(app != nil);
+    XCTAssert([app.testBundles count] == 5);
+    // Make sure we don't split when we don't want to
+    self.config.numSims = @4;
+    NSArray<BPXCTestFile *> *bundles = [BPPacker packTests:app.testBundles configuration:self.config andError:nil];
+    XCTAssert([bundles count] == [app.testBundles count]);
+    XCTAssert([[[bundles objectAtIndex:0] name] isEqualToString:@"BPSampleAppTests"]);
+    XCTAssert([[[bundles objectAtIndex:1] name] isEqualToString:@"BPSampleAppHangingTests"]);
+    XCTAssert([[[bundles objectAtIndex:2] name] isEqualToString:@"BPSampleAppCrashingTests"]);
+    XCTAssert([[[bundles objectAtIndex:3] name] isEqualToString:@"BPAppNegativeTests"]);
+    XCTAssert([[[bundles objectAtIndex:4] name] isEqualToString:@"BPSampleAppFatalErrorTests"]);
```

Comment:
I think this test is a little fragile in the sense that if we add tests to the app, it'll become flaky.

How about testing that the tests are ordered by time instead of testing a specific ordering?

Reply:
Done.